compiler/rustc_passes/src/check_attr.rs

```diff
@@ -1373,6 +1373,28 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
         }
     }
 
+    /// Checks if `#[may_dangle]` is applied to a lifetime or type generic parameter in `Drop` impl.
+    fn check_may_dangle(&self, hir_id: HirId, attr: &Attribute, target: Target) {
+        if matches!(target, Target::GenericParam(_))
+            && let hir::Node::GenericParam(param) = self.tcx.hir_node(hir_id)
+            && matches!(
+                param.kind,
+                hir::GenericParamKind::Lifetime { .. } | hir::GenericParamKind::Type { .. }
+            )
+            && matches!(param.source, hir::GenericParamSource::Generics)
+            && let parent_hir_id = self.tcx.parent_hir_id(hir_id)
+            && let hir::Node::Item(item) = self.tcx.hir_node(parent_hir_id)
+            && let hir::ItemKind::Impl(impl_) = item.kind
+            && let Some(trait_) = impl_.of_trait
```

Comment:
```suggestion
            && let hir::Node::Item(hir::Item { kind: hir::ItemKind::Impl(impl_), .. }) = self.tcx.hir_node(parent_hir_id)
            && let Some(trait_) = impl_.of_trait
```

Reply:
This leads to somewhat ugly formatting:

```rust
&& let hir::Node::Item(hir::Item { kind: hir::ItemKind::Impl(impl_), .. }) =
    self.tcx.hir_node(parent_hir_id)
```

I think the current version is nicer, but I don’t feel that strong about it, so I can make this change if you prefer it.